File: Validate Parenthesis.js
//validate parathesis

function isValid(s) {
    const stack = [];
    const map = {
        ')': '(', 
        '}': '{',
        ']': '['
    };

    for (const char of s) {
        // If closing bracket
        if (char in map) {
            // Pop and compare
            if (stack.pop() !== map[char]) {
                return false;
            }
        } 
        // Opening bracket
        else {
            stack.push(char);
        }
    }

    // Stack must be empty at the end
    return stack.length === 0;
}

File: Validate Parenthesis.py
def is_valid(s):
    stack = []
    mapping = {
        ')': '(',
        '}': '{',
        ']': '['
    }

    for char in s:
        # If closing bracket
        if char in mapping:
            # Stack empty or mismatch
            if not stack or stack.pop() != mapping[char]:
                return False
        else:
            # Opening bracket
            stack.append(char)

    # Stack should be empty at the end
    return len(stack) == 0


# User input
if __name__ == "__main__":
    s = input("Enter brackets string: ")
    print(is_valid(s))

